yateto/codegen/gemm/gemmgen.py

```diff
@@ -74,13 +84,94 @@ def generate(self, cpp, routineCache):
       flops = 2 * m.size() * n.size() * k.size()
     
     if isinstance(self._gemm_cfg, BLASlike):
```

Comment:
There is a way which came to my mind. It is based on double-dispatch (something like the Visitor pattern). Each *if-block* goes to a separate *generate* function. Selection of a specific *generate* function is delegated of sub-classes of *GemmTool* class (which self._gemm_cfg belongs to). Of course something needs to be done with the code above that you mentioned.

I am still not sure whether the double-dispatch is a good idea. What do you think? I would suggest to tackle this problem after the merge.

Reply:
Double-dispatch is fine here. This is wrong, btw: num_rows=d.leftTerm.memoryLayout._bbox[0].stop,

Reply:
Gemmforge uses a different format for describing matrices i.e., num_rows is total number of rows including padding. The same is for num_columns. There is check in yateto which ensures that all tensor operations are for tensors with ranks less or equal to 2 in case of GPU computing.

The next big release will learn gemmforge to operate on tensors and combine multiple tensors sub-expressions in one GPU kernel